### ISSUE TITLE
Asynchronous step verification

### ISSUE THREAD

**Issue by methodus** [stepstone-tech/android-material-stepper#89]
Hi,

I was wondering how to implement an asynchronous step verification. I have a process which requires the user to enter special credentials before he can proceed. If the credentials are wrong, the step verification should fail. Well, I could add yet another button to perform the verification before the user clicked on next, but I would like to have it done by the already existing interaction flow.

As the verifiyStep is a synchronous call on the main thread, I cannot use network calls.

It is also not possible to return to the previous step if the BlockedStep returns an error, for instance when the database is locked or some temporary network error occurred. In this case, I have to check it in the next step and return to the previous one, which breaks the independence between two separate steps.

Or did I missed something?

Thank you for any response.

**Comment by zawadz88** [comment]
Hi @methodus,
I think this is related to #68.
E.g. you could use a `BlockingStep` to make the verification request with entered credentials and if it fails show an error.

As for:

> It is also not possible to return to the previous step if the BlockedStep returns an error, for instance when the database is locked or some temporary network error occurred. In this case, I have to check it in the next step and return to the previous one, which breaks the independence between two separate steps.

I'm not sure what do you mean TBH... You can go to any step by calling `StepperLayout#setCurrentStepPosition(int)`. Also, if the operation in `BlockingStep` fails you could simply stay on the current step. 
Please see Delayed Transition in the sample app.

**Comment by methodus** [comment]
Yeah, after reading and understanding the functionality of BlockedSteps I came to this approach:

verifyStep just checks if the required input fields are filled before the user proceeds.

Then, I implemented the onNextClicked callbacks for BlockedStep which performs the required network requests. If they succeed, I call goToNextStep(). If it fails, I call callback.getStepperLayout().setErrorState(true) to update the Tab bar layout. It will also show a local error message what failed.

I didn't know, that if I don't call goToNextStep() the process stays on the current step.

So, all good.
